pkg/occlient/occlient.go

```diff
@@ -377,43 +391,145 @@ func (c *Client) GetImageStreamsNames(namespace string) ([]string, error) {
+	if imageNS == "" {
+
+		// First try finding imagestream from current namespace
+		currentNSImageStream, e := c.imageClient.ImageStreams(currentProjectName).Get(imageName, metav1.GetOptions{})
+		if e != nil {
+			err = errors.Wrapf(e, "no match found for : %s in namespace %s", imageName, currentProjectName)
+		} else {
+			if isTagInImageStream(*currentNSImageStream, imageTag) {
+				return currentNSImageStream, nil
+			}
+		}
+
+		// If not in current namespace, try finding imagestream from openshift namespace
+		openshiftNSImageStream, e := c.imageClient.ImageStreams(OpenShiftNameSpace).Get(imageName, metav1.GetOptions{})
+		if e != nil {
+			// The image is not available in current Namespace.
+			err = errors.Wrapf(e, "%s\n.no match found for : %s in namespace %s", err.Error(), imageName, OpenShiftNameSpace)
+		} else {
+			if isTagInImageStream(*openshiftNSImageStream, imageTag) {
+				return openshiftNSImageStream, nil
+			}
+		}
+		if e != nil && err != nil {
+			// Imagestream not found in openshift and current namespaces
+			return nil, err
+		}
+
+		// Required tag not in openshift and current namespaces
+		return nil, errors.Errorf("image stream %s with tag %s not found in openshift and %s namespaces", imageName, imageTag, currentProjectName)
+
+	} else {
+
+		// Fetch imagestream from requested namespace
+		imageStream, err = c.imageClient.ImageStreams(imageNS).Get(imageName, metav1.GetOptions{})
+		if err != nil {
+			return nil, errors.Wrapf(
+				err, "no match found for %s in namespace %s", imageName, imageNS,
+			)
+		}
+		if !isTagInImageStream(*imageStream, imageTag) {
+			return nil, errors.Errorf("image stream %s with tag %s not found in %s namespaces", imageName, imageTag, currentProjectName)
+		}
+	}
+
+	return imageStream, nil
+}
+
+// GetExposedPorts retruns image namespace and list of ContainerPorts that are exposed by given image
+func (c *Client) GetExposedPorts(imageNS string, imageName string, imageTag string) (string, []corev1.ContainerPort, error) {
 	var containerPorts []corev1.ContainerPort
+	var err error
+	var imageStream *imagev1.ImageStream
 
 	log.Debugf("Checking for exact match of builderImage with ImageStream")
-	imageStream, err := c.imageClient.ImageStreams(OpenShiftNameSpace).Get(imageName, metav1.GetOptions{})
+
+	imageStream, err = c.GetImageStream(imageNS, imageName, imageTag)
 	if err != nil {
-		log.Debugf("No exact match found: %s", err.Error())
-		return nil, errors.Wrapf(err, "unable to find matching builder image %s", imageName)
-	} else {
-		tagFound := false
-		for _, tag := range imageStream.Status.Tags {
-			// look for matching tag
-			if tag.Tag == imageTag {
-				tagFound = true
-				log.Debugf("Found exact image tag match for %s:%s", imageName, imageTag)
-				// ImageStream holds tag history
-				// first item is the latest one
+		errors.Wrapf(err, "error fetching exposed ports for image %s:%s", imageName, imageTag)
+	}
+	imageNS = imageStream.ObjectMeta.Namespace
```

Comment:
I just don't understand it :-(
Here you are setting `imageNS` to the namespace of the object that you got at line 478 (`imageStream, err = c.GetImageStream(imageNS, imageName, imageTag)`), in that call you specify `imageNS` so `imageStream.ObjectMeta.Namespace` will already be the same namespace as you `imageNS` because you are requesting ImageStream from `imageNS` namespace.

Reply:
OK, I'm starting to understand `c.GetImageStream(imageNS, imageName, imageTag)` works even when `imageNS` is empty. 
To be honest, it is a bit weird :-| We need a lot of comments to make all this logic clear.

Reply:
Ack